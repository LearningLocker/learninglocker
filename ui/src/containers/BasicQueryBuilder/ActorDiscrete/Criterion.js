import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Map, Set, List } from 'immutable';
import { connect } from 'react-redux';
import QueryBuilderAutoComplete from 'ui/components/AutoComplete2/QueryBuilderAutoComplete';
import { CriterionOperator, CriterionValue, CriterionWrapper } from 'ui/containers/BasicQueryBuilder/styled';
import Operator from '../Operator';

class Criterion extends Component {
  static propTypes = {
    section: PropTypes.instanceOf(Map),
    criterion: PropTypes.instanceOf(Map),
    filter: PropTypes.instanceOf(Map),
    onCriterionChange: PropTypes.func,
    onDeleteCriterion: PropTypes.func,
  }

  static defaultProps = {
    filter: new Map(),
  }

  state = {
    tempOperator: 'In'
  }

  shouldComponentUpdate = ({ section, criterion, filter }, { tempOperator }) => !(
    this.props.section.equals(section) &&
    this.props.criterion.equals(criterion) &&
    this.props.filter.equals(filter) &&
    this.state.tempOperator === tempOperator
  );

  canDeleteCriterion = () =>
    this.props.onDeleteCriterion !== undefined

  getSearchStringToFilter = () =>
    this.props.section.get('searchStringToFilter')

  getOptionQuery = option =>
    this.props.section.get('getModelQuery')(option)

  getOptionDisplay = option =>
    this.props.section.get('getModelDisplay')(option)

  getOptionIdentifier = option =>
    this.props.section.get('getModelIdent')(option)

  getQueryOption = query =>
    this.props.section.get('getQueryModel')(query)

  getCriterionQuery = (operator, criterion) => {
    switch (operator) {
      case 'Out': return new Map({ $nor: criterion });
      default: return new Map({ $or: criterion });
    }
  }

  getValues = () => {
    if (!this.canDeleteCriterion()) return new List();
    const operator = this.getOperator();
    if (operator === 'Out') return this.props.criterion.get('$nor', new List());
    return this.props.criterion.get('$or', new List());
  }

  getOperator = () => {
    if (!this.canDeleteCriterion()) return this.state.tempOperator;
    if (this.props.criterion.has('$nor')) return 'Out';
    return 'In';
  }

  changeCriterion = (operator, values) =>
    this.props.onCriterionChange(new Map({
      $comment: this.props.criterion.get('$comment'),
    }).merge(this.getCriterionQuery(operator, values)))

  changeValues = (values) => {
    const canDeleteCriterion = values.size === 0 && this.canDeleteCriterion();
    if (canDeleteCriterion) {
      return this.props.onDeleteCriterion();
    }
    return this.changeCriterion(this.getOperator(), values);
  }

  onAddOption = (model) => {
    if (!model.isEmpty()) {
      const values = this.getValues();
      const newValue = this.getOptionQuery(model);
      this.changeValues(values.push(newValue));
    }
  }

  onRemoveOption = (model) => {
    const values = this.getValues();
    const newValue = this.getOptionQuery(model);

    this.changeValues(
      values.filter(value => !value.equals(newValue))
    );
  }

  changeOperator = (operator) => {
    if (!this.canDeleteCriterion()) {
      return this.setState({ tempOperator: operator });
    }
    return this.changeCriterion(operator, this.getValues());
  }

  render = () => (
    <CriterionWrapper>
      <CriterionOperator>
        <Operator
          operators={new Set(['In', 'Out'])}
          operator={this.getOperator()}
          onOperatorChange={this.changeOperator} />
      </CriterionOperator>
      <CriterionValue isFullWidth>
        <QueryBuilderAutoComplete
          values={this.getValues()
            .map(this.getQueryOption)}
          filter={this.props.filter}
          schema={this.props.schema}
          selectOption={this.onAddOption}
          deselectOption={this.onRemoveOption}
          parseOption={this.getOptionDisplay}
          searchStringToFilter={this.getSearchStringToFilter()}
          parseOptionTooltip={this.getOptionIdentifier} />
      </CriterionValue>
    </CriterionWrapper>
  );
}

export default connect((_, ownProps) => {
  if (ownProps.section.has('sourceSchema')) {
    const schema = ownProps.section.get('sourceSchema');
    return { schema };
  }
  const path = ownProps.section.get('keyPath');
  const filter = new Map({ path: new Map({ $eq: path.join('.') }) });
  return { schema: 'querybuildercachevalue', filter };
}, {})(Criterion);
