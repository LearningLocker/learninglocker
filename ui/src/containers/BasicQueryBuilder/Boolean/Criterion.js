import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Map } from 'immutable';
import {
  CriterionButton,
  CriterionValue,
  CriterionWrapper,
  RadioLabel
} from 'ui/containers/BasicQueryBuilder/styled';
import { uniqueId } from 'lodash';

export class Criterion extends Component {
  static propTypes = {
    section: PropTypes.instanceOf(Map),
    criterion: PropTypes.instanceOf(Map),
    onCriterionChange: PropTypes.func,
    onDeleteCriterion: PropTypes.func,
  }

  shouldComponentUpdate = ({ section, criterion }) => !(
    this.props.section.equals(section) &&
    this.props.criterion.equals(criterion)
  );

  canDeleteCriterion = () =>
    this.props.onDeleteCriterion !== undefined;

  getValueQuery = value =>
    this.props.section.get('getValueQuery')(value);

  getKey = () => this.props.section.get('keyPath').join('.');

  getSubQuery = () => {
    const { criterion } = this.props;
    return criterion.get(this.getKey(), undefined);
  }

  getValue = () => {
    if (!this.canDeleteCriterion()) {
      return '';
    }

    return this.getSubQuery();
  }

  changeCriterion = value => this.props.onCriterionChange(new Map({
    $comment: this.props.criterion.get('$comment'),
  }).set(this.getKey(), value));

  changeValue = (value) => {
    const canDeleteCriterion = value === '' && this.canDeleteCriterion();
    if (canDeleteCriterion) {
      return this.props.onDeleteCriterion();
    }

    this.changeCriterion(value);
  }

  handleValueChange = (event) => {
    const value = event.target.value;

    if (value === '') {
      return this.changeValue(undefined);
    }

    if (value === 'true') {
      return this.changeValue(true);
    } else if (value === 'false') {
      return this.changeValue(false);
    }
  }

  render = () => {
    const canDeleteCriterion = this.canDeleteCriterion();
    const inputName = this.getKey();
    const trueId = uniqueId(`${inputName}-`);
    const falseId = uniqueId(`${inputName}-`);

    return (
      <CriterionWrapper>
        <CriterionValue isFullWidth={!canDeleteCriterion}>
          <div>
            <RadioLabel htmlFor={trueId}>
              True:&nbsp;&nbsp;
            </RadioLabel>
            <input
              id={trueId}
              type="radio"
              name={inputName}
              value="true"
              checked={this.getValue() === true}
              onChange={this.handleValueChange} />
          </div>
          <div>
            <RadioLabel htmlFor={falseId}>
              False:&nbsp;&nbsp;
            </RadioLabel>
            <input
              id={falseId}
              type="radio"
              name={inputName}
              value="false"
              checked={this.getValue() === false}
              onChange={this.handleValueChange} />
          </div>
        </CriterionValue>
        {(canDeleteCriterion &&
          <CriterionButton
            className={'btn btn-default btn-xs'}
            onClick={this.props.onDeleteCriterion} >
            <i className="ion-minus-round" />
          </CriterionButton>
        )}
      </CriterionWrapper>
    );
  }
}

export default Criterion;
