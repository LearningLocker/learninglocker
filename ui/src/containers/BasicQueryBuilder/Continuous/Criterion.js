import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Map, Set } from 'immutable';
import moment from 'moment-timezone';
import { toTimezone } from 'lib/constants/timezones';
import DatePicker from 'ui/components/Material/DatePicker';
import {
  CriterionButton,
  CriterionOperator,
  CriterionValue,
  CriterionWrapper
} from 'ui/containers/BasicQueryBuilder/styled';
import Operator from '../Operator';
import { symbolOpToMongoOp } from './helpers';

/**
 * @param {immutable.Map} section - {
 *                                    getQueryDisplay: (query: immutable.Map) => string,
 *                                    getValueQuery: (value: string) => immutable.Map,
 *                                    keyPath: immutable.List<string>,
 *                                  }
 * @param {immutable.Map} criterion - {
 *                                      $comment: string,
 *                                      [key:string]: immutable.Map,
 *                                    }
 * @param {(criterion: immutable.Map) => void} onCriterionChange
 * @param {() => void} onDeleteCriterion
 */
class Criterion extends Component {
  static propTypes = {
    timezone: PropTypes.string,
    orgTimezone: PropTypes.string.isRequired,
    section: PropTypes.instanceOf(Map).isRequired,
    criterion: PropTypes.instanceOf(Map).isRequired,
    onCriterionChange: PropTypes.func.isRequired,
    onDeleteCriterion: PropTypes.func.isRequired,
  };

  shouldComponentUpdate = nextProps => !(
    this.props.timezone === nextProps.timezone &&
    this.props.orgTimezone === nextProps.orgTimezone &&
    this.props.section.equals(nextProps.section) &&
    this.props.criterion.equals(nextProps.criterion)
  );

  /**
   * @param {immutable.Map} - query e.g. { $dte: '2019-02-03T09:00Z' }
   * @returns {string} value e.g. '2019-02-03T09:00Z'
   */
  getQueryDisplay = query => this.props.section.get('getQueryDisplay')(query);

  /**
   * @param {string} - value e.g. '2019-02-03T09:00Z'
   * @returns {string} query e.g. { $dte: '2019-02-03T09:00Z' }
   */
  getValueQuery = value => this.props.section.get('getValueQuery')(value);

  /**
   * @returns {string} key e.g. '$timestamp'
   */
  getKey = () =>
    this.props.section.get('keyPath').join('.');

  /**
   * @returns {immutable.Map} - query e.g. { $dte: '2019-02-03T09:00Z' }
   */
  getSubQuery = () =>
    this.props.criterion.get(this.getKey());

  /**
   * Get date object from query
   *
   * @returns {Date}
   */
  getDateValue = () =>
    moment(this.getValue(), 'YYYY-MM-DD').toDate();

  /**
   * Get date time value from query
   * @returns {string} - "{YYYY-MM-DD}T{HH:mm}Z"
   */
  getValue = () => {
    const operator = this.getOperator();
    const subQuery = this.getSubQuery();

    const mongoOp = symbolOpToMongoOp(operator);
    const queryValue = subQuery.get(mongoOp);
    return this.getQueryDisplay(queryValue);
  };

  /**
   * Get operator symbol from props or state
   *
   * @returns {string} symbol operator: "<", ">", "<=", or ">="
   */
  getOperator = () => {
    const subQuery = this.getSubQuery();
    if (subQuery.has('$gt')) return '>';
    if (subQuery.has('$lte')) return '<=';
    if (subQuery.has('$gte')) return '>=';
    return '<';
  };

  /**
   * @param {string} operator
   * @param {string} datetimeString - date time format "{YYYY-MM-DD}T{HH:mm}Z"
   */
  onChangeCriterion = (operator, datetimeString) => {
    const key = this.getKey();
    const mongoOp = symbolOpToMongoOp(operator);

    this.props.onCriterionChange(new Map({
      $comment: this.props.criterion.get('$comment'),
      [key]: new Map({ [mongoOp]: this.getValueQuery(datetimeString) }),
    }));
  };

  /**
   * @param {string} operator - "<", ">", "<=", or ">="
   */
  onChangeOperator = operator =>
    this.onChangeCriterion(operator, this.getValue());

  /**
   * @param {*} - argument of onChange in components/Material/DatePicker
   */
  onChangeDate = (value) => {
    const yyyymmdd = moment.parseZone(value).format('YYYY-MM-DD');
    const timezone = toTimezone(this.props.timezone || this.props.orgTimezone);
    const z = moment(yyyymmdd).tz(timezone).format('Z');
    this.onChangeCriterion(this.getOperator(), `${yyyymmdd}T00:00${z}`);
  };

  render = () => (
    <CriterionWrapper>
      <CriterionOperator>
        <Operator
          operators={new Set(['>', '<', '>=', '<='])}
          operator={this.getOperator()}
          onOperatorChange={this.onChangeOperator} />
      </CriterionOperator>

      <CriterionValue isFullWidth={false}>
        <DatePicker
          value={this.getDateValue()}
          onChange={this.onChangeDate} />
      </CriterionValue>

      <CriterionButton
        className={'btn btn-default btn-xs'}
        onClick={this.props.onDeleteCriterion}>
        <i className="ion-minus-round" />
      </CriterionButton>
    </CriterionWrapper>
  );
}

export default Criterion;
