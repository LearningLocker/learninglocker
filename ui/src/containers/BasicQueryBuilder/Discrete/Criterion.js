import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Map, Set, List } from 'immutable';
import { connect } from 'react-redux';
import QueryBuilderAutoComplete from
  'ui/components/AutoComplete2/QueryBuilderAutoComplete';
import { CriterionOperator, CriterionValue, CriterionWrapper } from 'ui/containers/BasicQueryBuilder/styled';
import Operator from '../Operator';
import { opToString, stringToOp, Operators } from './helpers';


class Criterion extends Component {
  static propTypes = {
    section: PropTypes.instanceOf(Map),
    criterion: PropTypes.instanceOf(Map),
    filter: PropTypes.instanceOf(Map),
    onCriterionChange: PropTypes.func,
    onDeleteCriterion: PropTypes.func,
  }

  static defaultProps = {
    filter: new Map(),
  }

  shouldComponentUpdate = ({ section, criterion, filter }) => !(
    this.props.section.equals(section) &&
    this.props.criterion.equals(criterion) &&
    this.props.filter.equals(filter)
  );

  /**
   * @returns {string}
   */
  getQueryKey = () =>
    this.props.section.get('getQueryKey', '')

  getSearchStringToFilter = () =>
    this.props.section.get('searchStringToFilter')

  getOptionQuery = option =>
    this.props.section.get('getModelQuery')(option)

  getOptionDisplay = option =>
    this.props.section.get('getModelDisplay')(option)

  getOptionIdentifier = option =>
    this.props.section.get('getModelIdent')(option)

  getQueryOption = query =>
    this.props.section.get('getQueryModel')(query)

  /**
   * @returns {Operators}
   */
  getOperator = () => {
    const key = this.getQueryKey();
    const subQuery = this.props.criterion.get(key, new Map());
    return subQuery.has(Operators.IN) ? Operators.IN : Operators.NIN;
  }

  /**
   * @returns {immutable.Set}
   */
  getValues = () => {
    const key = this.getQueryKey();
    const subQuery = this.props.criterion.get(key, new Map());
    const operator = this.getOperator();
    return subQuery.get(operator, new List()).toSet();
  }

  /**
   * @param {Operators} operator
   * @param {immutable.Set} values
   * @returns {void}
   */
  onChangeCriterion = (operator, values) => {
    const key = this.getQueryKey();
    this.props.onCriterionChange(new Map({
      $comment: this.props.criterion.get('$comment'),
      [key]: new Map({
        [operator]: values.toList()
      })
    }));
  }

  onAddOption = (model) => {
    if (model.isEmpty()) {
      return;
    }
    const newValue = this.getOptionQuery(model);
    const values = this.getValues();
    const added = values.add(newValue);
    this.onChangeCriterion(this.getOperator(), added);
  }

  onRemoveOption = (model) => {
    const values = this.getValues();
    const removingValue = this.getOptionQuery(model);
    const removed = values.remove(removingValue);

    if (removed.size === 0) {
      this.props.onDeleteCriterion();
      return;
    }

    this.onChangeCriterion(this.getOperator(), removed);
  }

  onChangeOperator = operator =>
    this.onChangeCriterion(operator, this.getValues());

  render = () => (
    <CriterionWrapper>
      <CriterionOperator>
        <Operator
          operators={new Set(['In', 'Out'])}
          operator={opToString(this.getOperator())}
          onOperatorChange={str => this.onChangeOperator(stringToOp(str))} />
      </CriterionOperator>
      <CriterionValue isFullWidth>
        <QueryBuilderAutoComplete
          values={this.getValues()
            .toList()
            .map(this.getQueryOption)}
          filter={this.props.filter}
          schema={this.props.schema}
          selectOption={this.onAddOption}
          deselectOption={this.onRemoveOption}
          parseOption={this.getOptionDisplay}
          searchStringToFilter={this.getSearchStringToFilter()}
          parseOptionTooltip={this.getOptionIdentifier} />
      </CriterionValue>
    </CriterionWrapper>
  );
}

export default connect((_, ownProps) => {
  if (ownProps.section.has('sourceSchema')) {
    const schema = ownProps.section.get('sourceSchema');
    return { schema };
  }
  const path = ownProps.section.get('keyPath');
  const filter = new Map({ path: new Map({ $eq: path.join('.') }) });
  return { schema: 'querybuildercachevalue', filter };
}, {})(Criterion);
