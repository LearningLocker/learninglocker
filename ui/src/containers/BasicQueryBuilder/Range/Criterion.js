import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Map, Set } from 'immutable';
import {
  CriterionButton,
  CriterionOperator,
  CriterionValue,
  CriterionWrapper
} from 'ui/containers/BasicQueryBuilder/styled';
import Operator from '../Operator';

export class Criterion extends Component {
  static propTypes = {
    section: PropTypes.instanceOf(Map),
    criterion: PropTypes.instanceOf(Map),
    onCriterionChange: PropTypes.func,
    onDeleteCriterion: PropTypes.func,
  }

  state = {
    tempOperator: '>'
  }

  shouldComponentUpdate = ({ section, criterion }, { tempOperator }) => !(
    this.props.section.equals(section) &&
    this.props.criterion.equals(criterion) &&
    this.state.tempOperator === tempOperator
  );

  canDeleteCriterion = () =>
    this.props.onDeleteCriterion !== undefined;

  getValueQuery = value =>
    this.props.section.get('getValueQuery')(value);

  getKey = () => this.props.section.get('keyPath').join('.');

  getSubQuery = () => {
    const { criterion } = this.props;
    return criterion.get(this.getKey(), this.getCriterionSubQuery('>', ''));
  }
  getQueryOption = query =>
    this.props.section.get('getQueryModel')(query)

  getCriterionSubQuery = (operator, value) => {
    switch (operator) {
      case '>': return new Map({ $gt: this.getValueQuery(value) });
      case '>=': return new Map({ $gte: this.getValueQuery(value) });
      case '<=': return new Map({ $lte: this.getValueQuery(value) });
      case '==': return new Map({ $eq: this.getValueQuery(value) });
      case '!=': return new Map({ $ne: this.getValueQuery(value) });
      default: return new Map({ $lt: this.getValueQuery(value) });
    }
  }

  getCriterionQuery = (operator, value) => {
    const key = this.getKey();
    return new Map({
      [key]: this.getCriterionSubQuery(operator, value)
    });
  }

  getValue = () => {
    if (!this.canDeleteCriterion()) return '';
    const operator = this.getOperator();
    const subQuery = this.getSubQuery();
    let queryValue;

    if (operator === '>') queryValue = subQuery.get('$gt');
    else if (operator === '>=') queryValue = subQuery.get('$gte');
    else if (operator === '<=') queryValue = subQuery.get('$lte');
    else if (operator === '!=') queryValue = subQuery.get('$ne');
    else if (operator === '==') queryValue = subQuery.get('$eq');
    else queryValue = subQuery.get('$lt');

    return queryValue;
  }

  getOperator = () => {
    if (!this.canDeleteCriterion()) return this.state.tempOperator;
    const subQuery = this.getSubQuery();
    if (subQuery.has('$gt')) return '>';
    if (subQuery.has('$lte')) return '<=';
    if (subQuery.has('$gte')) return '>=';
    if (subQuery.has('$eq')) return '==';
    if (subQuery.has('$ne')) return '!=';
    return '<';
  }

  getMax = () => this.props.section.get('max', 1);

  getStep = () => this.props.section.get('step', 0.1);

  changeCriterion = (operator, value) => this.props.onCriterionChange(new Map({
    $comment: this.props.criterion.get('$comment'),
  }).merge(this.getCriterionQuery(operator, value)));

  changeValue = (value) => {
    const canDeleteCriterion = value === '' && this.canDeleteCriterion();
    if (canDeleteCriterion) {
      return this.props.onDeleteCriterion();
    }

    this.changeCriterion(this.getOperator(), value);
  }

  changeOperator = (operator) => {
    if (!this.canDeleteCriterion()) {
      return this.setState({ tempOperator: operator });
    }
    this.changeCriterion(operator, this.getValue());
  }

  handleValueChange = (event) => {
    if (event.target.value === '') {
      return this.changeValue('');
    }
    const value = event.target.value * 1;
    return this.changeValue(value);
  }

  render = () => {
    const canDeleteCriterion = this.canDeleteCriterion();

    return (
      <CriterionWrapper>
        <CriterionOperator>
          <Operator
            operators={new Set(['>', '<', '>=', '<=', '==', '!='])}
            operator={this.getOperator()}
            onOperatorChange={this.changeOperator} />
        </CriterionOperator>
        <CriterionValue isFullWidth={!canDeleteCriterion}>
          <input
            type="number"
            min="0"
            max={this.getMax()}
            step={this.getStep()}
            value={this.getValue()}
            onChange={this.handleValueChange} />
        </CriterionValue>
        {(canDeleteCriterion &&
          <CriterionButton
            className={'btn btn-default btn-xs'}
            onClick={this.props.onDeleteCriterion}>
            <i className="ion-minus-round" />
          </CriterionButton>
        )}
      </CriterionWrapper>
    );
  }
}

export default Criterion;
