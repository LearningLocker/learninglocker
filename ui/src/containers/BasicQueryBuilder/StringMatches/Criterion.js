import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { List, Map, Set } from 'immutable';
import { connect } from 'react-redux';
import MultiInput from 'ui/components/MultiInput';
import { CriterionOperator, CriterionValue, CriterionWrapper } from 'ui/containers/BasicQueryBuilder/styled';
import Operator from '../Operator';

const Operators = {
  IN: '$in',
  NIN: '$nin'
};

const opToString = op => ((op === Operators.IN) ? 'In' : 'Out');
const stringToOp = str => ((str === 'In') ? Operators.IN : Operators.NIN);

/**
 * Criterion Component
 *
 * expected structure of criterion
 * type keyPathType = string;
 *
 * Map ({
 *   '$comment': string
 *   [keyPathType]: Map ({
 *     [Operator]: List<string>
 *   })
 * })
 */
class Criterion extends Component {
  static propTypes = {
    section: PropTypes.instanceOf(Map),
    criterion: PropTypes.instanceOf(Map),
    onCriterionChange: PropTypes.func,
    onDeleteCriterion: PropTypes.func,
  }

  state = {
    tempOperator: Operators.IN
  }

  shouldComponentUpdate = ({ section, criterion }, { tempOperator }) => !(
    this.props.section.equals(section) &&
    this.props.criterion.equals(criterion) &&
    this.state.tempOperator === tempOperator
  );

  /**
   * @returns {boolean}
   */
  hasQuery = () =>
    this.props.onDeleteCriterion !== undefined

  /**
   * @returns {string}
   */
  getKeyPath = () =>
    this.props.section.get('keyPath', []).join('.')

  /**
   * @returns {immutable.Set<string>}
   */
  getValues = () => {
    if (!this.hasQuery()) return new Set();

    const keyPath = this.getKeyPath();
    const opValues = this.props.criterion.get(keyPath, new Map());
    const operator = opValues.has(Operators.IN) ? Operators.IN : Operators.NIN;
    return opValues.get(operator, new List()).toSet();
  }

  /**
   * @returns {Operators}
   */
  getOperator = () => {
    if (!this.hasQuery()) return this.state.tempOperator;

    const keyPath = this.getKeyPath();
    const opValues = this.props.criterion.get(keyPath, new Map());
    return opValues.has(Operators.IN) ? Operators.IN : Operators.NIN;
  }

  /**
   * @param {Operators} operator
   * @param {immutable.Set<string>} values
   * @returns {void}
   */
  onChangeCriterion = (operator, values) => {
    const keyPath = this.getKeyPath();
    this.props.onCriterionChange(new Map({
      $comment: this.props.criterion.get('$comment'),
      [keyPath]: new Map({
        [operator]: values.toList()
      })
    }));
  }

  /**
   * @param {immutable.Set<string>} values
   * @returns {void}
   */
  onChangeValues = (values) => {
    if (values.size === 0 && this.hasQuery()) {
      this.props.onDeleteCriterion();
      return;
    }
    this.onChangeCriterion(this.getOperator(), values);
  }

  /**
   * @param {string} value
   * @returns {void}
   */
  onAddValue = (value) => {
    const values = this.getValues();
    this.onChangeValues(values.add(value));
  }

  /**
   * @param {string} value
   * @returns {void}
   */
  onRemoveValue = (value) => {
    const values = this.getValues();
    this.onChangeValues(values.delete(value));
  }

  /**
   * @param {Operators} operator
   * @returns {void}
   */
  onChangeOperator = (operator) => {
    if (!this.hasQuery()) {
      this.setState({ tempOperator: operator });
      return;
    }
    this.onChangeCriterion(operator, this.getValues());
  }

  render = () => (
    <CriterionWrapper>
      <CriterionOperator>
        <Operator
          operators={new Set(['In', 'Out'])}
          operator={opToString(this.getOperator())}
          onOperatorChange={str => this.onChangeOperator(stringToOp(str))} />
      </CriterionOperator>
      <CriterionValue isFullWidth>
        <MultiInput
          options={this.getValues()
            .toArray()}
          onAddOption={this.onAddValue}
          onRemoveOption={this.onRemoveValue} />
      </CriterionValue>
    </CriterionWrapper>
  );
}

export default connect((_, ownProps) => {
  const path = ownProps.section.get('keyPath');
  const filter = new Map({ path: new Map({ $eq: path.join('.') }) });
  return { schema: 'querybuildercachevalue', filter };
}, {})(Criterion);
