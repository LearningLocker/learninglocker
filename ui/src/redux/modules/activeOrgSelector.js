import { createSelector } from 'reselect';
import { modelsSchemaIdSelector } from 'ui/redux/selectors';
import { activeOrgIdSelector } from 'ui/redux/modules/router';

const activeOrgSelector = createSelector(
  [state => state, activeOrgIdSelector],
  (state, activeOrgId) =>
    modelsSchemaIdSelector('organisation', activeOrgId)(state),
);

export default activeOrgSelector;
