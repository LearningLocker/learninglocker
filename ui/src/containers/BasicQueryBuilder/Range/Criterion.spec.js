import React from 'react';
import 'jest-styled-components';
import renderer from 'react-test-renderer';
import { Map } from 'immutable';
import { initialSections } from 'ui/redux/modules/queryBuilder';
import Criterion from './Criterion.js';

describe('BasicQueryBuilder -> Range Criterion', () => {
  test('render', () => {
    const section = initialSections.getIn(['result', 'children', 'scaled']);
    const criterion = renderer
      .create(<Criterion section={section} />)
      .toJSON();

    expect(criterion).toMatchSnapshot();
  });

  test('with value', () => {
    const section = initialSections.getIn(['result', 'children', 'scaled']);

    const mockCriteria = new Map({
      'statement.result.score.scaled': new Map({
        $gt: 0.6
      })
    });

    const criterion = renderer
      .create(<Criterion criterion={mockCriteria} section={section} />)
      .toJSON();

    expect(criterion).toMatchSnapshot();
  });

  test('operator change', () => {
    const mockMerge = jest.fn();

    const onCriterionChange = jest.fn().mockReturnValueOnce({
      merge: mockMerge
    });

    const testCriterion = new Criterion({
      criterion: new Map({
        $comment: 'A comment',
      }),
      section: initialSections.getIn(['result', 'children', 'scaled']),
      onCriterionChange,
      onDeleteCriterion: jest.fn()
    });


    testCriterion.changeOperator('<=');

    expect(onCriterionChange.mock.calls.length).toEqual(1);
    expect(onCriterionChange.mock.calls[0][0].get('statement.result.score.scaled')).toEqual(new Map({ $lte: '' }));
  });

  test('value change', () => {
    const mockMerge = jest.fn();

    const onCriterionChange = jest.fn().mockReturnValueOnce({
      merge: mockMerge
    });

    const testCriterion = new Criterion({
      criterion: new Map({
        $comment: 'A comment',
      }),
      section: initialSections.getIn(['result', 'children', 'scaled']),
      onCriterionChange,
      onDeleteCriterion: jest.fn()
    });


    testCriterion.handleValueChange({
      target: {
        value: '0.5'
      }
    });

    expect(onCriterionChange.mock.calls.length).toEqual(1);
    expect(onCriterionChange.mock.calls[0][0].get('statement.result.score.scaled')).toEqual(new Map({ $gt: 0.5 }));
  });
});
