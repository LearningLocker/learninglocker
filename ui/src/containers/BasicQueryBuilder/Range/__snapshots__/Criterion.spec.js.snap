// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BasicQueryBuilder -> Range Criterion render 1`] = `
.c2 {
  position: relative;
  cursor: pointer;
}

.c2:focus {
  outline: 0;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c4 {
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

.c5 {
  margin-left: 2px;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  width: 10px;
}

.c6 {
  min-height: 30px;
  width: 100%;
}

.c0 {
  margin-bottom: 4px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c1 {
  padding-top: 2px;
  width: 15%;
  margin-right: 2%;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  min-height: 30px;
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    <div
      className="c2"
      onBlur={[Function]}
      onFocus={[Function]}
      tabIndex={0}
    >
      <div
        className="c3 btn btn-default btn-xs"
      >
        <div
          className="c4"
        >
          &gt;
        </div>
        <div
          className="c5"
        >
          <div
            className="ion-arrow-up-b"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="c6"
  >
    <input
      max={1}
      min="0"
      onChange={[Function]}
      step={0.1}
      type="number"
      value=""
    />
  </div>
</div>
`;

exports[`BasicQueryBuilder -> Range Criterion with value 1`] = `
.c2 {
  position: relative;
  cursor: pointer;
}

.c2:focus {
  outline: 0;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c4 {
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

.c5 {
  margin-left: 2px;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  width: 10px;
}

.c6 {
  min-height: 30px;
  width: 100%;
}

.c0 {
  margin-bottom: 4px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c1 {
  padding-top: 2px;
  width: 15%;
  margin-right: 2%;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  min-height: 30px;
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    <div
      className="c2"
      onBlur={[Function]}
      onFocus={[Function]}
      tabIndex={0}
    >
      <div
        className="c3 btn btn-default btn-xs"
      >
        <div
          className="c4"
        >
          &gt;
        </div>
        <div
          className="c5"
        >
          <div
            className="ion-arrow-up-b"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="c6"
  >
    <input
      max={1}
      min="0"
      onChange={[Function]}
      step={0.1}
      type="number"
      value=""
    />
  </div>
</div>
`;
